modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java

```diff
@@ -142,39 +151,47 @@ public void onResponse(ReindexTaskState taskState) {
 
             @Override
             public void onFailure(Exception e) {
+                if (e instanceof VersionConflictEngineException) {
+                    if (isDone.compareAndSet(false, true)) {
```

Comment:
I think we should retry and maybe even re-read the doc to check if our allocation is still current and use the new term/seqno if so. This is unfortunately necessary until we get rid of dirty reads and ensure CAS'es only have positive failures. This can certainly be done in a follow-up (I realize this is not the focus of this PR).